### PR TITLE
[DROOLS-7196] Overloaded method resolution is different between mvel …

### PR DESCRIPTION
**Pull request by tkobayas**
…and executable model (varargs)

- Java doesn't coerce int to Long (Wrapper class). It's not a varargs issue
- Documented as known difference
- Added KnownExecModelDifferenceTest for this kind of behavior differences

**JIRA**:
https://issues.redhat.com/browse/DROOLS-7196

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] tests</b>

- for a <b>full downstream build</b> 
  - for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  - for <b>github actions</b> job: add the label `run_fdb`

- <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

- <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

- <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-main</b>

- for <b>quarkus lts checks</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins run quarkus-lts</b>

- for a <b>specific quarkus lts check</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-lts</b>

- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] native</b>

- for <b>mandrel checks</b>  
  Run native checks against Mandrel image
  Please add comment: <b>Jenkins run mandrel</b>

- for a <b>specific mandrel check</b>  
  Run native checks against Mandrel image  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] mandrel</b>

 - for <b>mandrel lts checks</b>  
  Run native checks against Mandrel image and quarkus lts branch
  Please add comment: <b>Jenkins run mandrel-lts</b>

- for a <b>specific mandrel lts check</b>  
  Run native checks against Mandrel image and quarkus lts branch
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] mandrel-lts</b>
</details>

### CODE REVIEW

**Comment by kie-ci3** [comment]
**(tests) - drools job** `#845` was: **FAILURE**
Possible explanation: Pipeline failure or project build failure

Please look here: https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/drools.tests.drools/845/display/redirect

**Test results:**
- PASSED: 20865
- FAILED: 0

Those are the test failures: none

See console log:
<details>
<summary><b>Console Logs</b></summary>
[Pipeline] tool<br/>[Pipeline] envVarsForTool<br/>[Pipeline] tool<br/>[Pipeline] envVarsForTool<br/>[Pipeline] withEnv<br/>[Pipeline] {<br/>[Pipeline] script<br/>[Pipeline] {<br/>[Pipeline] dir<br/>[2022-12-12T14:15:10.327Z] Running in /home/jenkins/workspace/KIE/kogito/main/pullrequest/drools.tests.drools/bc/kiegroup_drools/drools<br/>[Pipeline] {<br/>[Pipeline] withCredentials<br/>[2022-12-12T14:15:10.423Z] Masking supported pattern matches of $TOKEN<br/>[Pipeline] {<br/>[Pipeline] configFileProvider<br/>[2022-12-12T14:15:10.474Z] provisioning config files...<br/>[2022-12-12T14:15:10.490Z] copy managed file [kogito_pr_settings] to file:/home/jenkins/workspace/KIE/kogito/main/pullrequest/drools.tests.drools/bc/kiegroup_drools/drools/maven-settings.xml<br/>[Pipeline] {<br/>[Pipeline] }<br/>[Pipeline] // configFileProvider<br/>[Pipeline] }<br/>[Pipeline] // withCredentials<br/>[Pipeline] }<br/>[Pipeline] // dir<br/>[Pipeline] }<br/>[Pipeline] // script<br/>[Pipeline] }<br/>[Pipeline] // withEnv<br/>[Pipeline] }<br/>[Pipeline] // stage<br/>[Pipeline] stage<br/>[Pipeline] { (Upload artifacts to repository)<br/>Stage "Upload artifacts to repository" skipped due to earlier failure(s)<br/>[Pipeline] }<br/>[Pipeline] // stage<br/>[Pipeline] stage<br/>[Pipeline] { (check space after build)<br/>Stage "check space after build" skipped due to earlier failure(s)<br/>[Pipeline] }<br/>[Pipeline] // stage<br/>[Pipeline] stage<br/>[Pipeline] { (Declarative: Post Actions)<br/>[Pipeline] script<br/>[Pipeline] {<br/>[Pipeline] sh<br/>[2022-12-12T14:15:11.431Z] + wget --no-check-certificate -qO - https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/drools.tests.drools/845/api/json<br/>[Pipeline] readJSON<br/>[Pipeline] sh<br/>[2022-12-12T14:15:11.803Z] + wget --no-check-certificate -qO - https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/drools.tests.drools/845/consoleText<br/>[2022-12-12T14:15:11.803Z] + tail -n 50<br/>
</details>

**Comment by tkobayas** [comment]
DROOLS-7196 was raised as a varargs issue, but actually an issue about coercion from `int` to `Long`. non-exec-model (mvel) is tolerant, so it works without an error. However, it's invalid in Java, so exec-model results in a build error.

I put this issue in the docs "Migration from non-executable model to executable model" section as a known difference rather than fixing one side.

Also I added KnownExecModelDifferenceTest for this kind of behavior differences.

Please review, @mariofusco @pibizza @hellowdan @kaldesai , thanks!

**Review by mariofusco — APPROVED** [review]

**Review by hellowdan — APPROVED** [review]

File: drools-docs/src/modules/ROOT/pages/migration-guide/_nonexec-model-to-exec-model.adoc [hunk 1]
```diff
@@ -65,3 +65,30 @@ then
   }
 end
 ----
+
+=== Invalid coercion
+For example, Java allows coercion from `int` to `long`. However, Java doesn't allow coercion from `int` to `Long`. If you have a rule to call `setWrapperLong` method which accepts `Long`,
```
Comment (kaldesai) [inline drools-docs/src/modules/ROOT/pages/migration-guide/_nonexec-model-to-exec-model.adoc]:
```suggestion
For example, Java allows coercion from `int` to `long`. However, Java doesn't allow coercion from `int` to `Long`. If you have the rule to call the `setWrapperLong` method which accepts `Long`.
```

File: drools-docs/src/modules/ROOT/pages/migration-guide/_nonexec-model-to-exec-model.adoc [hunk 2]
```diff
@@ -65,3 +65,30 @@ then
...
+rule Rule1
+dialect "mvel"
+when
+  $f : Fact()
+then
+  $f.setWrapperLong(10);
+end
+----
+
+`non-executable model` coerces `10` to `Long`, so it doesn't throw an Exception. However, `executable model` throws a build error.
```
Comment (kaldesai) [inline drools-docs/src/modules/ROOT/pages/migration-guide/_nonexec-model-to-exec-model.adoc]:
```suggestion
`non-executable model` coerces `10` to `Long`, so it doesn't throw an Exception. However, the `executable model` throws a build error.
```

**Review by kaldesai — APPROVED** [review]
@tkobayas Minimal suggestions. Otherwise, your content looks good to me. Thank you!

**Comment by tkobayas** [comment]
@kaldesai Thank you! Applied the suggestions.

**Comment by kie-ci3** [comment]
**(tests) - kogito-apps job** `#833` was: **UNSTABLE**
Possible explanation: This should be test failures

Please look here: https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/drools.tests.downstream.kogito-apps/833/display/redirect

**Test results:**
- PASSED: 1960
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/drools.tests.downstream.kogito-apps/833/testReport/org.kie.kogito.it.jobs/ProcessAsyncIT/Build_projects___testAsync/">Build projects / org.kie.kogito.it.jobs.ProcessAsyncIT.testAsync</a></summary>
java.lang.AssertionError: <br/>1 expectation failed.<br/>JSON path hello doesn't match.<br/>Expected: null<br/>  Actual: Hello Tiago<br/><br/>
</details>

**Comment by tkobayas** [comment]
@mariofusco Please merge this PR, thanks!

**Pull request merged**